### PR TITLE
Fix empty checksum file/snapshot corruption

### LINKED ISSUE #14699: IllegalStateException: found snapshot 0. A previous snapshot is most likely corrupted
**Describe the bug**

During startup of the broker Zeebe encounters snapshot 0. Likely an indication of some corruption.

We encountered this before on prod. The solution was to delete the partition data and restart the broker. After this the broker recovers. When we first found it https://github.com/camunda/zeebe/issues/14486 was created. This was released in 8.3.0. However, today it occurred again on 8.3.0. 

**To Reproduce**

Unclear, but it's probably hard to do as you'd need to introduce some corruption.

**Expected behavior**

It is handled more gracefully and doesn't prevent the broker from reaching a bootloop.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
java.lang.IllegalStateException: In partition 1 expected to find a snapshot at index >= log's first index 30443521, but found snapshot 0. A previous snapshot is most likely corrupted.
at io.atomix.raft.utils.StateUtil.verifySnapshotLogConsistent ( io/atomix.raft.utils/StateUtil.java:33 )
at io.atomix.raft.impl.RaftContext.<init> ( io/atomix.raft.impl/RaftContext.java:223 )
at io.atomix.raft.impl.DefaultRaftServer$Builder.build ( io/atomix.raft.impl/DefaultRaftServer.java:242 )
at io.atomix.raft.impl.DefaultRaftServer$Builder.build ( io/atomix.raft.impl/DefaultRaftServer.java:216 )
at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer ( io/atomix.raft.partition.impl/RaftPartitionServer.java:159 )
at io.atomix.raft.partition.impl.RaftPartitionServer.<init> ( io/atomix.raft.partition.impl/RaftPartitionServer.java:91 )
at io.atomix.raft.partition.RaftPartition.createServer ( io/atomix.raft.partition/RaftPartition.java:127 )
at io.atomix.raft.partition.RaftPartition.initServer ( io/atomix.raft.partition/RaftPartition.java:110 )
at io.atomix.raft.partition.RaftPartition.bootstrap ( io/atomix.raft.partition/RaftPartition.java:94 )
at io.camunda.zeebe.broker.partitioning.startup.steps.RaftBootstrapStep.startup ( io/camunda.zeebe.broker.partitioning.startup.steps/RaftBootstrapStep.java:31 )
at io.camunda.zeebe.broker.partitioning.startup.steps.RaftBootstrapStep.startup ( io/camunda.zeebe.broker.partitioning.startup.steps/RaftBootstrapStep.java:14 )
at io.camunda.zeebe.scheduler.startup.StartupProcess.proceedWithStartupSynchronized ( io/camunda.zeebe.scheduler.startup/StartupProcess.java:161 )
at io.camunda.zeebe.scheduler.startup.StartupProcess.lambda$proceedWithStartupSynchronized$3 ( io/camunda.zeebe.scheduler.startup/StartupProcess.java:169 )
at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run ( io/camunda.zeebe.scheduler.future/FutureContinuationRunnable.java:28 )
at io.camunda.zeebe.scheduler.ActorJob.invoke ( io/camunda.zeebe.scheduler/ActorJob.java:94 )
at io.camunda.zeebe.scheduler.ActorJob.execute ( io/camunda.zeebe.scheduler/ActorJob.java:45 )
at io.camunda.zeebe.scheduler.ActorTask.execute ( io/camunda.zeebe.scheduler/ActorTask.java:119 )
at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask ( io/camunda.zeebe.scheduler/ActorThread.java:130 )
at io.camunda.zeebe.scheduler.ActorThread.doWork ( io/camunda.zeebe.scheduler/ActorThread.java:108 )
at io.camunda.zeebe.scheduler.ActorThread.run ( io/camunda.zeebe.scheduler/ActorThread.java:227 )
```

</p>
</details>

**Environment:**
- OS: SaaS
- Zeebe Version: 8.3.0

https://console.cloud.google.com/logs/query;query=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22camunda-saas-int-chaos%22%0Aresource.labels.namespace_name%3D%22f2275ddc-b3a0-44b4-bc8a-cdfc242514f1-zeebe%22%0Aresource.labels.location%3D%22europe-west1%22%0Aresource.labels.cluster_name%3D%22worker-chaos-1%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.pod_name%3D%22zeebe-0%22;pinnedLogId=2023-10-12T12:31:31.413159098Z%2Fa7opurgfpm3zpxtj;cursorTimestamp=2023-10-12T12:53:35.349727194Z;duration=PT3H?project=camunda-saas-int-chaos

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR fixes two possible cases where the checksum file may be empty.

1. **The node crashes after the file is created but before anything is written to it.** We now write to a temporary file (same file name, but with a ".tmp" suffix, e.g. `6513-1-3-12.checksum.tmp`), before atomically moving it to its correct location, and then flushing the directory file descriptor.
2. **The contents of the checksum file are not flushed to disk before restart**. This can happen since `FileOutputStream#flush` does absolutely nothing, and relies on the OS to decide when to flush. We now use a `FileChannel`, whose `force` method calls `fsync` directly, thus guaranteeing the contents are available.

This PR adds a new test utility, `STracer` (and `STracerAssert`), which allows you to trace system calls. Right now, there's only support for `fsync`, and no support to inject faults, but this can be expanded in the future.

## Related issues

closes #14699 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: test-util/src/main/java/io/camunda/zeebe/test/util/STracer.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  public static STracer traceFor(final Syscall syscall, final Path outputFile, final boolean debug)
+      throws IOException {
+    final var pid = ProcessHandle.current().pid();
+    final var output = outputFile.toAbsolutePath().toString();
+    final var builder =
+        new ProcessBuilder()
+            .command(
+                "strace",
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply test-util/src/main/java/io/camunda/zeebe/test/util/STracer.java:36]: Hm, OK, but does it matter all that much? I get the potential security risk, but what's the alternative that doesn't terribly suck? I can't predict all the places strace is installed on all the different distros :shrug:

[the remark this answers is not in the document]
Comment (oleschoenburg) [reply test-util/src/main/java/io/camunda/zeebe/test/util/STracer.java:36]: Yeah it's a test util, just dismiss the alert as such :shrug:

File: snapshot/src/test/java/io/camunda/zeebe/snapshots/impl/SnapshotChecksumTest.java [hunk 2]
```diff
@@ -123,8 +130,32 @@ public void shouldPersistChecksum() throws Exception {
     assertThat(actual.getCombinedValue()).isEqualTo(expectedChecksum.getCombinedValue());
   }
 
+  @DisabledIfEnvironmentVariable(named = "GITHUB_ACTIONS", matches = "true")
```
Comment (oleschoenburg) [inline snapshot/src/test/java/io/camunda/zeebe/snapshots/impl/SnapshotChecksumTest.java:133]: :upside_down_face: I think `CI` is the canonical environment variable set by all CI systems: https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables

File: test-util/src/main/java/io/camunda/zeebe/test/util/STracer.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+ */
+package io.camunda.zeebe.test.util;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.regex.Pattern;
+
+public final class STracer implements AutoCloseable {
```
Comment (oleschoenburg) [inline test-util/src/main/java/io/camunda/zeebe/test/util/STracer.java:16]: :+1: Very cool utility to have :rocket:

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java [hunk 4]
```diff
@@ -526,12 +524,15 @@ FileBasedSnapshot persistNewSnapshot(
...
-      // as a marker file to guarantee the move was complete and not partial
+      // it as a marker file to guarantee the move was complete and not partial
       final var destination = buildSnapshotDirectory(snapshotId);
       final var checksumPath = buildSnapshotsChecksumPath(snapshotId);
+      final var tmpChecksumPath =
+          checksumPath.resolveSibling(checksumPath.getFileName().toString() + TMP_CHECKSUM_SUFFIX);
       try {
-        SnapshotChecksum.persist(checksumPath, immutableChecksumsSFV);
+        SnapshotChecksum.persist(tmpChecksumPath, immutableChecksumsSFV);
+        Files.move(tmpChecksumPath, checksumPath, StandardCopyOption.ATOMIC_MOVE);
```
Comment (oleschoenburg) [inline snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java:534]: :x: I think we need to handle `AtomicMoveNotSupportedException` here. Just falling back to a non-atomic move is probably the only/best option?

**Review by lenaschoenburg — APPROVED** [review]
I haven't reviewed the new test utils in detail since they need to be fixed for CI anyway, but I tried the test locally.

I have one change request to ensure that we keep supporting systems where atomic file move is not supported. 

Feel free to merge without an additional review though.

File: snapshot/src/test/java/io/camunda/zeebe/snapshots/impl/SnapshotChecksumTest.java (hunk 2 above)
Comment (npepinpe) [reply snapshot/src/test/java/io/camunda/zeebe/snapshots/impl/SnapshotChecksumTest.java:133]: Good to know, but I wonder why the need for both :smile:

**Comment by npepinpe** [comment]
bors merge

**Comment by npepinpe** [comment]
Last commit was pushed on the wrong branch :facepalm:

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**